### DISCUSSION TITLE
lnd v0.16.3-beta

### DISCUSSION THREAD

**Discussion by Roasbeef** [lightningnetwork/lnd#7748] [Announcements]
This is the 3rd minor release in the 0.16.3 release. This release contains only bug fixes and is intended to optimize the recently added mempool watching logic, and also fix several suspected inadvertent force close vectors. 

# Database Migrations

This release contains no database migrations. 

# Verifying the Release

In order to verify the release, you'll need to have `gpg` or `gpg2` installed on your system. Once you've obtained a copy (and hopefully verified that as well), you'll first need to import the keys that have signed this release if you haven't done so already: 

```
curl https://raw.githubusercontent.com/lightningnetwork/lnd/master/scripts/keys/roasbeef.asc | gpg --import
```

Once you have the required PGP keys, you can verify the release (assuming `manifest-roasbeef-v0.16.3-beta.sig` and `manifest-v0.16.3-beta.txt` are in the current directory) with:

```
gpg --verify manifest-roasbeef-v0.16.3-beta.sig manifest-v0.16.3-beta.txt
```

You should see the following if the verification was successful:

```
gpg: Signature made Wed Sep 30 17:35:20 2020 PDT
gpg:                using RSA key 60A1FA7DA5BFF08BDCBBE7903BBD59E99B280306
gpg: Good signature from "Olaoluwa Osuntokun <laolu32@gmail.com>" [ultimate]
```

That will verify the signature of the manifest file, which ensures integrity and authenticity of the archive you've downloaded locally containing the binaries. Next, depending on your operating system, you should then re-compute the `sha256` hash of the archive with `shasum -a 256 <filename>`, compare it with the corresponding one in the manifest file, and ensure they match *exactly*.

## Verifying the Release Timestamp 

From this new version onwards, in addition time-stamping the _git tag_ with [OpenTimeStamps](https://opentimestamps.org/), we'll also now timestamp the manifest file along with its signature. Two new files are now included along with the rest of our release artifacts: ` manifest-roasbeef-v0.16.3-beta.txt.asc.ots`.

Assuming you have the opentimestamps client installed locally, the timestamps can be verified with the following commands: 
```
ots verify manifest-roasbeef-v0.16.3-beta.sig.ots -f manifest-roasbeef-v0.16.3-beta.sig
```

Alternatively, [the open timestamps website](https://opentimestamps.org/) can be used to verify timestamps if one doesn't have a `bitcoind` instance accessible locally. 

These timestamps should give users confidence in the integrity of this release even after the key that signed the release expires. 

## Verifying the Release Binaries

Our release binaries are fully reproducible. Third parties are able to verify that the release binaries were produced properly without having to trust the release manager(s). See our [reproducible builds guide](https://github.com/lightningnetwork/lnd/tree/master/build/release) for how this can be achieved.
The release binaries are compiled with `go1.20.3`, which is required by verifiers to arrive at the same ones.
They include the following build tags: `autopilotrpc`, `signrpc`, `walletrpc`, `chainrpc`, `invoicesrpc`, `neutrinorpc`, `routerrpc`, `watchtowerrpc`, `monitoring`, `peersrpc`, `kvdb_postrgres`, `kvdb_etcd` and `kvdb_sqlite`. Note that these are already included in the release script, so they do not need to be provided.

The `make release` command can be used to ensure one rebuilds with all the same flags used for the release. If one wishes to build for only a single platform, then `make release sys=<OS-ARCH> tag=<tag>` can be used. 

Finally, you can also verify the _tag_ itself with the following command:

```
$ git verify-tag v0.16.3-beta
gpg: Signature made Tue Sep 15 18:55:00 2020 PDT
gpg:                using RSA key 60A1FA7DA5BFF08BDCBBE7903BBD59E99B280306
gpg: Good signature from "Olaoluwa Osuntokun <laolu32@gmail.com>" [ultimate]
```

## Verifying the Docker Images

To verify the `lnd` and `lncli` binaries inside the docker images against the signed, reproducible release binaries, there is a verification script in the image that can be called (before starting the container for example):

```shell
$ docker run --rm --entrypoint="" lightninglabs/lnd:v0.16.3-beta /verify-install.sh v0.16.3-beta
$ OK=$?
$ if [ "$OK" -ne "0" ]; then echo "Verification failed!"; exit 1; done
$ docker run lightninglabs/lnd [command-line options]
```

# Building the Contained Release 

Users are able to rebuild the target release themselves without having to fetch any of the dependencies. In order to do so, assuming
that `vendor.tar.gz` and `lnd-source-v0.16.3-beta.tar.gz` are in the current directory, follow these steps:

```
tar -xvzf vendor.tar.gz
tar -xvzf lnd-source-v0.16.3-beta.tar.gz
GO111MODULE=on go install -v -mod=vendor -ldflags "-X github.com/lightningnetwork/lnd/build.Commit=v0.16.3-beta" ./cmd/lnd
GO111MODULE=on go install -v -mod=vendor -ldflags "-X github.com/lightningnetwork/lnd/build.Commit=v0.16.3-beta" ./cmd/lncli
```

The `-mod=vendor` flag tells the `go build` command that it doesn't need to fetch the dependencies, and instead, they're all enclosed in the local vendor directory. 

Additionally, it's now possible to use the [enclosed `release.sh` script to bundle a release for a _specific_ system like so](https://github.com/lightningnetwork/lnd/pull/2191): 

```
make release sys="linux-arm64 darwin-amd64"
```

⚡️⚡️⚡️ OK, now to the rest of the release notes! ⚡️⚡️⚡️

# Release Notes

https://github.com/lightningnetwork/lnd/blob/master/docs/release-notes/release-notes-0.16.3.md

# Contributors (Alphabetical Order)

* Elle Mouton
* Olaoluwa Osuntokun
* Yong Yu

<hr /><em>This discussion was created from the release <a href='https://github.com/lightningnetwork/lnd/releases/tag/v0.16.3-beta'>lnd v0.16.3-beta</a>.</em>

**Comment by schildbach** [comment]
Is is expected that

```
docker run --rm --entrypoint="" lightninglabs/lnd:v0.16.3-beta /verify-install.sh v0.16.3-beta
```

still yields:

```
ERROR: Not enough valid signatures found!
  Valid signatures found: 0
  Valid signatures required: 5
```

I would expect that 6 days after release at least one or two sigs to show up?

**Reply by guggero** [reply]
There are enough signatures (we don't publish the release otherwise), but something seems to be not working correctly in the docker version. It runs fine if I run it outside of docker. Let me investigate, thanks for the report.

**Reply by guggero** [reply]
I was able to track down the problem to a new behavior in the `gpg` version `v2.4.1` that is bundled in the Docker base image. See https://github.com/lightningnetwork/lnd/pull/7758

**Comment by schildbach** [comment]
The issue is still present in 0.16.4, so I've now reported it as a bug: #7809

**Reply by ellemouton** [reply]
@schildbach, yeah - the fix PR hasnt been merged yet
